scvi/metrics/log_likelihood.py

```diff
@@ -18,6 +24,71 @@ def compute_log_likelihood(vae, data_loader):
     return log_lkl / n_samples
 
 
+def compute_marginal_log_likelihood(vae, data_loader, n_samples_mc=100):
+    """ Computes a biased estimator for log p(x), which is the marginal log likelihood.
+        Despite its bias, the estimator still converges to the real value
+        of log p(x) when n_samples_mc (for Monte Carlo) goes to infinity
+        (a fairly high value like 100 should be enough)
+        Due to the Monte Carlo sampling, this method is not as computationally efficient
+        as computing only the reconstruction loss
+    """
+    # Uses MC sampling to compute a tighter lower bound
+    log_lkl = 0
+    for i_batch, tensors in enumerate(data_loader):
+        sample_batch, local_l_mean, local_l_var, batch_index, labels = tensors
+        batch_log_lkl = compute_tighter_log_likelihood_sample(vae, sample_batch, local_l_mean,
+                                                              local_l_var, batch_index,
+                                                              labels, n_samples_mc=n_samples_mc)
+        log_lkl += torch.sum(batch_log_lkl).item()
+    n_samples = (len(data_loader.dataset)
+                 if not (hasattr(data_loader, 'sampler') and hasattr(data_loader.sampler, 'indices')) else
+                 len(data_loader.sampler.indices))
+    return - log_lkl / n_samples
+
+
+def logsumexp(inputs, dim=None, keepdim=False):
+    """Numerically stable logsumexp.
+
+    Args:
+        inputs: A Variable with any shape.
+        dim: An integer.
+        keepdim: A boolean.
+
+    Returns:
+        Equivalent of log(sum(exp(inputs), dim=dim, keepdim=keepdim)).
+    """
+    # For a 1-D array x (any array along a single dimension),
+    # log sum exp(x) = s + log sum exp(x - s)
+    # with s = max(x) being a common choice.
+    if dim is None:
+        inputs = inputs.view(-1)
+        dim = 0
+    s, _ = torch.max(inputs, dim=dim, keepdim=True)
+    outputs = s + (inputs - s).exp().sum(dim=dim, keepdim=True).log()
+    if not keepdim:
+        outputs = outputs.squeeze(dim)
+    return outputs
+
+
+def compute_tighter_log_likelihood_sample(vae, sample_batch, local_l_mean, local_l_var, batch_index,
+                                          labels, n_samples_mc=100):
```

Comment:
It'd be better to merge put this code into `compute_marginal_log_likelihood` rather than making it it's own function, since that's the only place it's called